General information for rpc spec generation (#52)
smartdevicelink/sdl_overview_guides: docs/RPC Spec Generator/index.md

# RPC Spec Generator

The RPC Spec repository contains a parsing script that is used by the following platform's generator implementations:

- SDL Core
- SDL IOS
- SDL Java Suite
- SDL JavaScript

The RPC Spec repository is imported as a git submodule into each of the previously mentioned platforms. The RPC Spec parser reads the MOBILE_API.xml which is the specification for all RPC messages communicated between a proxy platform and SDL Core. Each platform contains a code generator for creating RPC classes and files. These generators greatly reduce the amount of effort for adding new RPC messages.

## Using the RPC Spec Generator

All developers who are using an SDL platform should run these commands in the SDL platform source directory to ensure the latest RPC spec and parser is being used in their builds:

```
git submodule init
git submodule update
```

Using the parser requires a minimum python version, 3.5. The rest of the dependencies for the generators can be installed via:

```

> python3 -m pip install -r InterfaceParser/requirements.txt

Author: I was referring to the path in the rpc_spec repository.

Example:
```xml
<enum deprecated="boolean" internal_scope="string" name="string" platform="string" since="string" until="string">
  <!--Zero or more repetitions:-->
  <description>string</description>
  <!--Zero or more repetitions:-->
 <element hexvalue="string" internal_name="string" name="string" rootscreen="boolean" since="string" value="integer">
    <!--Zero or more repetitions:-->
    <description>string</description>
    <!--Zero or more repetitions:-->
    <warning>string</warning>
  </element>
  <!--Optional:-->
  <history>
    <enum/>
  </history>
</enum>
```

SDL has two different enum types: `string` and `integer`. If `"hexvalue"` or `"value"` attribute exists, the enum type is `integer`, otherwise the enum type is `string`.

The `<element>` has a required `"name"` attribute. For `string` enums, the value of `"name"` attribute will be one of the possible values of the particular `<enum>`.  For `integer` enums, the value of `"value"` attribute will be one of the possible values. The `"hexvalue"` is just a hexadecimal representation of the `"value"` attribute and either hexvalue or value can be used.

> `<element>` could also have an `"internal_name"` attribute. This attribute is not used in communication with SDL Core, but it describes the `<element>` name for SDL libraries.

Author: Added example

# RPC Spec Generator

> The RPC Spec repository contains a parsing script that is used by the following platform's generator implementations:

Reviewer: ```suggestion
The RPC Spec repository contains a parsing script that is used by the following platforms' generator implementations:
```

- SDL Core

> - SDL IOS

Reviewer: ```suggestion
- SDL iOS
```

- SDL Java Suite

> - SDL JavaScript

Reviewer: ```suggestion
- SDL JavaScript Suite
```

The RPC Spec repository is imported as a git submodule into each of the previously mentioned platforms. The RPC Spec parser reads the MOBILE_API.xml which is the specification for all RPC messages communicated between a proxy platform and SDL Core. Each platform contains a code generator for creating RPC classes and files. These generators greatly reduce the amount of effort for adding new RPC messages.

## Using the RPC Spec Generator

All developers who are using an SDL platform should run these commands in the SDL platform source directory to ensure the latest RPC spec and parser is being used in their builds:

```
git submodule init
git submodule update
```

Using the parser requires a minimum python version, 3.5. The rest of the dependencies for the generators can be installed via:

```
python3 -m pip install -r InterfaceParser/requirements.txt
```

> Each platform has their own generator implementation but the instructions for generating their respective RPC code is the same. Refer to the following command line options for running a SDL platform generator script:

Reviewer: ```suggestion
Each platform has their own generator implementation but the instructions for generating their respective RPC code is the same. Refer to the following command line options for running an SDL platform generator script:
```

Example:
```xml
<enum deprecated="boolean" internal_scope="string" name="string" platform="string" since="string" until="string">
  <!--Zero or more repetitions:-->
  <description>string</description>
  <!--Zero or more repetitions:-->
 <element hexvalue="string" internal_name="string" name="string" rootscreen="boolean" since="string" value="integer">
    <!--Zero or more repetitions:-->
    <description>string</description>
    <!--Zero or more repetitions:-->
    <warning>string</warning>
  </element>
  <!--Optional:-->
  <history>
    <enum/>
  </history>
</enum>
```

SDL has two different enum types: `string` and `integer`. If `"hexvalue"` or `"value"` attribute exists, the enum type is `integer`, otherwise the enum type is `string`.

The `<element>` has a required `"name"` attribute. For `string` enums, the value of `"name"` attribute will be one of the possible values of the particular `<enum>`.  For `integer` enums, the value of `"value"` attribute will be one of the possible values. The `"hexvalue"` is just a hexadecimal representation of the `"value"` attribute and either hexvalue or value can be used.

> `<element>` could also have an `"internal_name"` attribute. This attribute is not used in communication with SDL Core, but it describes the `<element>` developer facing elemtn value. For example refer to the `HMILevel` enum:

Reviewer: ```suggestion
`<element>` could also have an `"internal_name"` attribute. This attribute is not used in communication with SDL Core, but it describes the `<element>` developer facing element value. For example refer to the `HMILevel` enum:
```

        <element name="NONE" internal_name="HMI_NONE" />
    </enum>
```

The HMILevel communicated to the proxies from SDL Core will use the `name` value (`FULL`, `NONE`, etc). However the `HMILevel` enum usage in SDL Core's code will use the `internal_name` value (`HMI_FULL`, `HMI_NONE`, etc).

### struct
`<struct>` is a complex data type. `<struct>` contains any number of `<param>`. The `<struct>` has a required `"name"` attribute.

Example:
```xml
<struct deprecated="boolean" name="string" since="string" until="string">
  <!--Optional:-->
  <history>
    <struct/>
  </history>
  <!--Zero or more repetitions:-->
  <description>string</description>
  <!--Zero or more repetitions:-->
  <param array="boolean" defvalue="integer|decimal|boolean|string" deprecated="boolean" mandatory="boolean" maxlength="integer" maxsize="integer" maxvalue="decimal" minlength="integer" minsize="integer" minvalue="decimal" name="string" since="string" type="string" until="string">
    <!--Zero or more repetitions:-->
    <description>string</description>
    <!--Optional:-->
    <history>
      <param/>
    </history>
  </param>
</struct>
```

> Each `<param>` requires `"name"`, `"type"` and `"mandatory"` attributes. The `"type"` attribute value should be one of `"Boolean"`, `"Float"`, `"Integer"`, `"String"` or one of the `<enum>`, `<struct>` names specified in the Mobile API.

Reviewer: ```suggestion
Each `<param>` requires `"name"`, `"type"`, and `"mandatory"` attributes. The `"type"` attribute value should be one of `"Boolean"`, `"Float"`, `"Integer"`, `"String"`, or one of the `<enum>`, `<struct>` names specified in the Mobile API.
```

`<param>` could have `"array"` attribute to represent an array of values or objects of the described type. Attributes `"maxsize"` and `"minsize"` provide additional restrictions to an array.

Numeric types can be restricted using `"minvalue"` and `"maxvalue"`. 

The `"defvalue"` attribute contains default value with different type, depends on `<param>` type, note: this attribute is not allowed for `<struct>` type.

### function

> The `<function>` element represents a specific RPC and message type of the Mobile API. It contains any number of `<param>`. The `<function>` has a required `"name"`, `"messagetype"`, `"functionID"` attributes.

Reviewer: ```suggestion
The `<function>` element represents a specific RPC and message type of the Mobile API. It contains any number of `<param>`. The `<function>` has required `"name"`, `"messagetype"`, and `"functionID"` attributes.
```

Example:
```xml

<function deprecated="boolean" functionID="string" messagetype="string" name="string" since="string" until="string">
  <!--Optional:-->
  <history>
    <function/>
  </history>
  <!--Zero or more repetitions:-->
  <description>string</description>
  <!--Zero or more repetitions:-->
  <param array="boolean" defvalue="integer|decimal|boolean|string" deprecated="boolean" mandatory="boolean" maxlength="integer" maxsize="integer" maxvalue="decimal" minlength="integer" minsize="integer" minvalue="decimal" name="string" platform="string" since="string" type="string" until="string">
    <!--Zero or more repetitions:-->
    <description>string</description>
    <!--Optional:-->
    <history>
      <param/>
    </history>
    <!--Zero or more repetitions:-->

>     <todo>string</todo>

Reviewer: Are these `todo`s expected?

Author: Yes the "todo" tag is technically supported. :(

    <!--Zero or more repetitions:-->
    <element name="string">
      <!--Optional:-->
      <description>string</description>
    </element>
  </param>
</function>
```
The `"messagetype"` attribute value should be either `"request"`, `"response"`, or `"notification"`. 

The `"functionID"` attribute value should match the `"name`" attribute of one `<element>` from the `"FunctionID"` `<enum>`.

> Just like `<param>` elements in `<struct>`, each `<param>` has required `"name"`, `"type"`, and `"mandatory"` attributes. The `"type"` attribute value should be one of `"Boolean"`, `"Float"`, `"Integer"`, `"String"` or the one of `<enum>`, `<struct>` name exists in XML.

Reviewer: ```suggestion
Just like `<param>` elements in `<struct>`, each `<param>` has required `"name"`, `"type"`, and `"mandatory"` attributes. The `"type"` attribute value should be one of `"Boolean"`, `"Float"`, `"Integer"`, `"String"`, or one of the `<enum>`, `<struct>` name exists in XML.
```

`<param>` could have `"array"` attribute which means the param represents array of described types. Attributes `"max*"` and `"min*"` provide additional restrictions. 

The `"defvalue"` attribute contains default value with different type, depends on `<param>` type, note: this attribute is not allowed for `<struct>` type.

### RPC Spec Versioning

> Any element, param, function, or struct can use following xml attributes to express what RPC spec version an item was added, removed, deprecated, or changed.

Reviewer: ```suggestion
Any element, param, function, or struct can use the following xml attributes to express in which RPC spec version an item was added, removed, deprecated, or changed.
```
